src/pages/achievement/control/subcomponents/editorTools/EditableAchievementCard.tsx

```diff
@@ -0,0 +1,221 @@
+import EditableAchievementBackground from './editableUtils/EditableAchievementBackground';
+import EditableAchievementDate from './editableUtils/EditableAchievementDate';
+import EditableAchievementTitle from './editableUtils/EditableAchievementTitle';
+import EditableAchievementGoals from './editableUtils/goals/EditableAchievementGoals';
+import EditableAchievementView from './editableView/EditableAchievementView';
+
+type EditableAchievementCardProps = {
+  achievement: AchievementItem;
+  inferencer: AchievementInferencer;
+  updateAchievements: any;
+  editAchievement: any;
+  forceRender: any;
+  adderId: number;
+  setAdderId: any;
+  addUnsavedChange: any;
+  removeUnsavedChange: any;
+  removeGoal: any;
+  removeAchievement: any;
+};
+
+function EditableAchievementCard(props: EditableAchievementCardProps) {
+  const {
+    achievement,
+    inferencer,
+    updateAchievements,
+    editAchievement,
+    adderId,
+    setAdderId,
+    addUnsavedChange,
+    removeUnsavedChange,
+    removeGoal,
+    removeAchievement
+  } = props;
+
+  const [editableAchievement, setEditableAchievement] = useState<AchievementItem>(achievement);
+  const { id, title, ability, deadline, cardTileUrl, release, goals } = editableAchievement;
+
+  const [hasChanges, setHasChanges] = useState<boolean>(false);
+  const [pendingUpload, setPendingUpload] = useState<boolean>(false);
+
+  const setUnsaved = () => {
+    if (!hasChanges) {
+      setHasChanges(true);
+      addUnsavedChange();
+    }
+  };
+
+  const setUpload = () => {
+    setPendingUpload(false);
+    removeUnsavedChange();
+  };
+
+  const handleSaveChanges = () => {
+    inferencer.modifyAchievement(editableAchievement);
+    setHasChanges(false);
+    setPendingUpload(true);
+
+    if (id === adderId) {
+      setAdderId(-1);
+    }
+  };
+
+  const handleUploadChanges = () => {
+    editAchievement(editableAchievement);
+    setUpload();
+  };
+
+  const handleDeleteAchievement = () => {
+    inferencer.removeAchievement(id);
+
+    removeAchievement(editableAchievement);
+    updateAchievements();
+
+    if (id === adderId) {
+      setAdderId(-1);
+    }
+  };
+
+  const handleDiscardChanges = () => {
+    setEditableAchievement(achievement);
+    setHasChanges(false);
+    setUpload();
+  };
+
+  const handleChangeTitle = (title: string) => {
+    setEditableAchievement({
+      ...editableAchievement,
+      title: title
+    });
+    setUnsaved();
+  };
+
+  const handleEditGoals = (goals: AchievementGoal[], shouldUpdate: boolean) => {
+    setEditableAchievement({
+      ...editableAchievement,
+      goals: goals
+    });
+
+    inferencer.modifyAchievement(editableAchievement);
+    editAchievement(editableAchievement);
+  };
+
+  const handleRemoveGoal = (goal: AchievementGoal) => {
+    removeGoal(goal, editableAchievement);
+  };
+
+  const handleChangeBackground = (cardTileUrl: string) => {
+    setEditableAchievement({
```

Comment:
Perhaps there is some way to reduce the repetitive code here using Hooks?

https://rangle.io/blog/simplifying-controlled-inputs-with-hooks/

Reply:
Ps ps, I think the code is fine cos each text field is interactive.